README.md
# pocket-sockets

A powerful and smooth sockets library for browser and _Node.js_, supporting both _WebSockets_ and regular _TCP_ sockets, encrypted or not.

This project is extracted as a standalone project from the [Universe-ai/core-js](https://github.com/universe-ai/core-js) project and can be used on its own.

## WebSockets vs. regular TCP sockets
_WebSockets_ are great when using a browser, however plain _TCP_ sockets are faster and a good choice if no one of the peers is a browser.

The overall interface for _pocket-sockets_ _WebSocket_ and _TCP_ sockets are the same so it is easy to switch between the underlying implementations.

## License
This project is released under the _MIT_ license. Refer to the [LICENSE](https://github.com/universe-ai/core-js/blob/master/LICENSE) file for details.


CHANGELOG.md
# CHANGELOG: pocket-sockets

## [1.1.0] - 20210304
Adds socket get functions to AbstractClient; fixes websocket protocol to work correctly in the browser and resolves bug when buffering events prior to handler getting registered.

## [1.0.1] - 20210223
Exposes the public API allowing usage as Node.js imported module.

## [1.0.0] - 20210222
First release.


test/WSClient.test.js
const assert = require("assert");

jest.mock("tls");
const tls = require("tls");

jest.mock("net");
const net = require("net");

const WSClient = require("../WSClient");

describe("WSClient", () => {
    describe("_socketConnect", () => {
        test("successful call setting USE_TLS", () => {
            assert.doesNotThrow(() => {
                const client = new WSClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": true,
                    "rejectUnauthorized": null
                });
                assert(client.connectOptions.host == "host.com");
                assert(client.connectOptions.port == 99);
                assert(client.connectOptions.secure == true);
                assert(client.connectOptions.rejectUnauthorized == null);

                assert(client.socket == null);
                client._socketConnect();
                assert(client.socket);
                assert(client.socket.onopen);
            });
        });

        test("successful call without USE_TLS", () => {
            assert.doesNotThrow(() => {
                const client = new WSClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                assert(client.connectOptions.host == "host.com");
                assert(client.connectOptions.port == 99);
                assert(client.connectOptions.secure == false);
                assert(client.connectOptions.rejectUnauthorized == null);

                assert(client.socket == null);
                client._socketConnect();
                assert(client.socket);
                assert(client.socket.onopen);
            });
        });
    });

    describe("_socketHook", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new WSClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                client._socketConnect();
                client._socketHook();
            });
        });
    });

    describe("_socketSend", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new WSClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                client._socketConnect();
                client.socket = {
                    "send": function(buffer, options) {
                        assert(buffer.toString() == "testdata123");
                        assert(Buffer.isBuffer(buffer));
                        assert(options.binary == true);
                        assert(options.compress == false);
                    }
                };
                client._socketSend(Buffer.from("testdata123"));
            });
        });
    });

    describe("_socketDisconnect", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new WSClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                client._socketConnect();
                let hasClosed = false;
                client.socket = {
                    "close": function() {
                        hasClosed = true;
                    }
                };
                assert(hasClosed == false);
                client._socketDisconnect();
                assert(hasClosed == true);
            });
        });
    });
});


test/WSServer.test.js
const assert = require("assert");

jest.mock("https");
const https = require("https");

jest.mock("http");
const http = require("http");

jest.mock("ws");
const WebSocket = require("ws");

const WSServer = require("../WSServer");
const WSClient = require("../WSClient");

describe("WSServer", () => {
    describe("constructor", () => {
        test("missing listenOptions", () => {
            assert.throws(() => {
                const server = new WSServer();
            }, /Bad options/);
        });

        test("successful call with cert", () => {
            assert.doesNotThrow(() => {
                https.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "secureConnection" || name == "error" || name == "close" || name == "tlsClientError");
                        assert(typeof fn == "function");
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null,
                    "cert": "valid-certificate"
                });
                assert(server.wsServer == null);
                assert(server.SocketType == WSClient);
            });
        });

        test("successful call without TLS", () => {
            assert.doesNotThrow(() => {
                http.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "connection" || name == "error" || name == "close" || name == "tlsClientError");
                        assert(typeof fn == "function");
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null
                });
                assert(server.wsServer == null);
                assert(server.SocketType == WSClient);
            });
        });
    });

    describe("_serverCreate", () => {
        test("successful call with USE_TLS", () => {
            assert.doesNotThrow(() => {
                https.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "secureConnection" || name == "error" || name == "close" || name == "tlsClientError");
                        assert(typeof fn == "function");
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null,
                    "cert": "valid-certificate"
                });
                assert(server.server);
            });
        });

        test("successful call without USE_TLS", () => {
            assert.doesNotThrow(() => {
                http.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "connection" || name == "error" || name == "close" || name == "tlsClientError");
                        assert(typeof fn == "function");
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null
                });
                assert(server.server);
            });
        });
    });

    describe("_serverListen", () => {
        test("overwritten server after object creation", () => {
            assert.throws(() => {
                https.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "secureConnection" || name == "error" || name == "close" || name == "tlsClientError");
                        assert(typeof fn == "function");
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null,
                    "cert": "valid-certificate"
                });
                server.server = null;
                server._serverListen();
            });
        });

        test("successful call", () => {
            assert.doesNotThrow(() => {
                https.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "connection" || name == "error" || name == "close" || name == "tlsClientError" || name == "listening" || name == "upgrade");
                        assert(typeof fn == "function");
                    },
                    "listen": function(options) {
                        assert(options.host == "host.com");
                        assert(options.port == 99);
                        assert(!options.ipv6Only); 
                    }
                });

                WebSocket.Server = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "connection" || name == "error" || name == "close"); 
                        assert(typeof fn == "function");
                    }
                });

                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null,
                    "cert": "valid-certificate"
                });
                server._serverListen();
            });
        });
    });

    describe("_serverClose", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                https.createServer = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                    },
                    "listen": function(options) {
                    }
                });
                WebSocket.Server = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                    }
                });
                const server = new WSServer({
                    "host": "host.com",
                    "port": 99,
                    "rejectUnauthorized": null,
                    "cert": "valid-certificate"
                });
                server._serverListen();

                let callCount = 0;
                server.server.close = function() {
                    callCount++;
                }
                server.wsServer.close = function() {
                    callCount++;
                }
                assert(callCount == 0);
                server._serverClose();
                assert(callCount == 2);
            });
        });
    });
});


test/TCPClient.test.js
const assert = require("assert");

jest.mock("tls");
const tls = require("tls");

jest.mock("net");
const net = require("net");

const TCPClient = require("../TCPClient");

describe("TCPClient", () => {
    describe("_socketConnect", () => {
        test("successful call setting USE_TLS", () => {
            assert.doesNotThrow(() => {
                const client = new TCPClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": true,
                    "rejectUnauthorized": null
                });
                assert(client.connectOptions.host == "host.com");
                assert(client.connectOptions.port == 99);
                assert(client.connectOptions.secure == true);
                assert(client.connectOptions.rejectUnauthorized == null);

                assert(client.socket == null);
                assert(client.isDisconnected == false);
                assert(client.bufferData == true);

                tls.connect = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "secureConnect");
                        assert(typeof fn == "function");
                    }
                });
                client._socketConnect();
            });
        });

        test("successful call without USE_TLS", () => {
            assert.doesNotThrow(() => {
                const client = new TCPClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                assert(client.connectOptions.host == "host.com");
                assert(client.connectOptions.port == 99);
                assert(client.connectOptions.secure == false);
                assert(client.connectOptions.rejectUnauthorized == null);

                assert(client.socket == null);
                assert(client.isDisconnected == false);
                assert(client.bufferData == true);

                net.connect = jest.fn().mockReturnValue({
                    "on": function(name, fn) {
                        assert(name == "connect");
                        assert(typeof fn == "function");
                    }
                });
                client._socketConnect();
            });
        });
    });

    describe("_socketHook", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new TCPClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                let counter = 0;
                client.socket = {
                    "on": function(name, fn) {
                        counter++;
                        assert(name == "data" || name == "error" || name == "close" );
                        assert(typeof fn == "function");
                    }
                };
                assert(counter == 0);
                client._socketHook();
                assert(counter == 3);
            });
        });
    });

    describe("_socketSend", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new TCPClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                client.socket = {
                    "write": function(buffer) {
                        assert(buffer.toString() == "testdata123");
                        assert(Buffer.isBuffer(buffer));
                    }
                };
                client._socketSend(Buffer.from("testdata123"));
            });
        });
    });

    describe("_socketDisconnect", () => {
        test("successful call", () => {
            assert.doesNotThrow(() => {
                const client = new TCPClient({
                    "host": "host.com",
                    "port": 99,
                    "secure": false,
                    "rejectUnauthorized": null
                });
                let hasEnded = false;
                client.socket = {
                    "end": function() {
                        hasEnded = true;
                    }
                };
                assert(hasEnded == false);
                client._socketDisconnect();
                assert(hasEnded == true);
            });
        });
    });
});


WSClient.js
// Add browser/browserify/parcel dependency
if(typeof "process" === "undefined" && !process && !process.versions && !process.versions.node) {
    require("regenerator-runtime/runtime");
}

let isBrowser;
let ws = null;
if (typeof WebSocket !== "undefined") {
    isBrowser = true;
    ws = WebSocket;
} else if (typeof MozWebSocket !== "undefined") {
    isBrowser = true;
    ws = MozWebSocket;
} else if (typeof window !== "undefined") {
    isBrowser = true;
    ws = window.WebSocket || window.MozWebSocket;
} else {
    isBrowser = false;
    ws = require("ws");
}

const AbstractClient = require("./AbstractClient.js");

/**
 * WebSocket client implementation.
 */
class WSClient extends AbstractClient
{
    /**
     * Specifies how the socket gets initialized and created, then establishes a connection.
     */
    _socketConnect()
    {
        const USE_TLS = this.connectOptions.secure ? true: false;

        let address;
        if(USE_TLS) {
            address = `wss://${this.connectOptions.host}:${this.connectOptions.port}`;
        }
        else {
            address = `ws://${this.connectOptions.host}:${this.connectOptions.port}`;
        }

        if(isBrowser) {
            this.socket = new ws(address);
            // Make sure binary type is set to ArrayBuffer instead of Blob
            this.socket.binaryType = "arraybuffer";
        } else {
            this.socket = new ws(address, {
                // Client certificate, can be set when using TLS, could be required by the server.
                cert: this.connectOptions.cert,
                key: this.connectOptions.key,

                // Certs are always fetched from the server,
                // If set to true we do not allow server self signed certs.
                rejectUnauthorized: this.connectOptions.rejectUnauthorized,

                // If validating the server cert (rejectUnauthorized == true)
                // then we can set the CA cert here which can make a self signed server cert
                // valid in our context.
                ca: this.connectOptions.ca,

                perMessageDeflate: false,
                maxPayload: 100 * 1024 * 1024,
            });
        }

        this.socket.onopen      = this._connect;
    }

    /**
     * Specifies hooks to be called as part of the connect procedure.
     */
    _socketHook()
    {
        this.socket.onmessage   = (msg) => {
            let data = msg.data;

            // Under Browser settings, convert message data from ArrayBuffer to Buffer.
            if (isBrowser) {
                const bytes = new Uint8Array(data);
                data = Buffer.from(bytes);
            }

            this._data(data);          // Incoming data
        };
        this.socket.onerror     = this._error;           // Error connecting
        this.socket.onclose     = this._disconnect;      // Socket closed
    }

    /**
     * Defines how data gets written to the socket.
     * @param {Buffer} buffer - data to be sent
     */
    _socketSend(buffer)
    {
        this.socket.send(buffer, {binary: true, compress: false});
    }

    /**
     * Defines the steps to be performed during disconnect.
     */
    _socketDisconnect()
    {
        this.socket.close();
    }
}

module.exports = WSClient;


TCPClient.js
const net = require("net");
const tls = require("tls");

const AbstractClient = require("./AbstractClient.js");

/**
 * TCP client socket implementation.
 */
class TCPClient extends AbstractClient
{

    /**
     * Specifies how the socket gets initialized and created, then establishes a connection.
     */
    _socketConnect()
    {
        const USE_TLS = this.connectOptions.secure ? true: false;

        if(USE_TLS) {
            const tlsOptions = {
                host: this.connectOptions.host,
                port: this.connectOptions.port,

                // Client certificate, must be set when using TLS if the server requires it.
                cert: this.connectOptions.cert,
                key: this.connectOptions.key,

                // Enforce server connection verification.
                // If this is set to true, then any requested certificate will be validated.
                rejectUnauthorized: this.connectOptions.rejectUnauthorized,

                // If validating the server cert (rejectUnauthorized == true)
                // then we can set the CA cert here which can make a self signed client cert valid in our context.
                ca: this.connectOptions.ca
            };
            this.socket = tls.connect(tlsOptions);
            this.socket.on("secureConnect", this._connect);
        }
        else {
            this.socket = net.connect({
                host: this.connectOptions.host,
                port: this.connectOptions.port,
            });
            this.socket.on("connect", this._connect);
        }
    }

    /**
     * Specifies hooks to be called as part of the connect procedure.
     */
    _socketHook()
    {
        this.socket.on("data", this._data);             // Incoming data
        this.socket.on("error", this._error);           // Error connecting
        this.socket.on("close", this._disconnect);      // Socket closed
    }

    /**
     * Defines how data gets written to the socket.
     * @param {Buffer} buffer - data to be sent
     */
    _socketSend(buffer)
    {
        this.socket.write(buffer);
    }

    /**
     * Defines the steps to be performed during disconnect.
     */
    _socketDisconnect()
    {
        this.socket.end();
    }
}

module.exports = TCPClient;


TCPServer.js
const net = require("net");
const tls = require("tls");

const TCPClient = require("./TCPClient");
const AbstractServer = require("./AbstractServer");

/**
 * TCP server implementation.
 */
class TCPServer extends AbstractServer
{
    constructor(listenOptions)
    {
        super(listenOptions);
        this.SocketType = TCPClient;
    }

    /**
     * Specifies how the server gets initialized, then creates the server with the specified options.
     */
    _serverCreate()
    {
        const USE_TLS = this.listenOptions.cert != null;

        if(USE_TLS) {
            const tlsOptions = {
                // Server certificate, must be set when using TLS.
                cert: this.listenOptions.cert,
                key: this.listenOptions.key,

                // Enforce client-side certification.
                // If this is set the server will request a client certificate.
                requestCert: this.listenOptions.requestCert,

                // Enforce client connection verification.
                // If this is set to true, then any requested certificate will be validated if requestCert is set.
                rejectUnauthorized: this.listenOptions.rejectUnauthorized,

                // If requesting and validating the client cert (requestCert == true, rejectUnauthorized == true)
                // then we can set the CA cert here which can make a self signed client cert valid in our context.
                ca: this.listenOptions.ca,

                // Milliseconds until we timeout a TLS-handshake.
                // This is important to have when a regular TCP socket is connecting to a TLS socket,
                // and the client is waiting for the server to send first message, then no handshake error will happen because no data is transmitted.
                handshakeTimeout: 30000,
            };
            this.server = tls.createServer(tlsOptions);
            this.server.on("secureConnection", this._connection);
        }
        else {
            this.server = net.createServer();
            this.server.on("connection", this._connection);
        }

        this.server.on("error", this._error);
        this.server.on("close", this._close);
        this.server.on("tlsClientError", (_err, socket) => socket.end());
    }

    /**
     * Starts a previously created server listening for connections.
     * Assumes the server is instantiated during object creation.
     */
    _serverListen()
    {
        this.server.listen({
            host: this.listenOptions.host,
            port: this.listenOptions.port,
            ipv6Only: this.listenOptions.ipv6Only,
        });
    }
}

module.exports = TCPServer;


VirtualClient.js
const AbstractClient = require("./AbstractClient.js");

class VirtualClient extends AbstractClient
{
    /**
     * @constructor
     * @param {VirtualSocket} [pairedSocket] When creating the second socket of a socket-pair provide the first socket as argument to get them paired.
     */
    constructor(pairedSocket)
    {
        super(null, {});

        this.pairedSocket = pairedSocket;

        /** We can set this to simulate some latency in the paired socket communication */
        this.latency = 0;  // Milliseconds

        /**
         * Queue of outgoing messages.
         * We need this if we use simulated latency,
         * because the ordering of setTimeout might not be guaranteed
         * for identical timeout values.
         */
        this.outQueue = [];

        /* Complete the pair by assigning this socket to the paired socket */
        if (this.pairedSocket) {
            this.pairedSocket.pairedSocket = this;
        }
    }

    /**
     * Set a simulated latency of the socket communications.
     *
     * @param {number} latency in milliseconds for each send
     */
    setLatency(latency)
    {
        this.latency = latency;
    }

    /**
     * Hook events on the socket.
     */
    _socketHook()
    {
        // We handle events in different ways since this is not an actual socket.
    }

    /**
     * Send the given buffer on socket.
     * @param {Buffer} buffer
     */
    _socketSend(buffer)
    {
        // Put msg into paired socket.
        if (this.pairedSocket) {
            this.outQueue.push(buffer);
            if (this.latency > 0) {
                setTimeout( () => this._copyToPaired(), this.latency);
            } else {
                this._copyToPaired();
            }
        }
    }

    /**
     * Specify the paired disconnect procedure.
     */
    _socketDisconnect()
    {
        if (this.pairedSocket) {
            this.pairedSocket._disconnect();
        }
        this._disconnect();
    }

    /**
     * Internal function to copy one message in the out queue to the paired socket.
     *
     */
    _copyToPaired()
    {
        const buffer = this.outQueue.shift();
        this.pairedSocket._data(buffer);
    }
}

function CreatePair()
{
    const socket1 = new VirtualClient();
    const socket2 = new VirtualClient(socket1);
    return [socket1, socket2];
}

module.exports = CreatePair;


WSServer.js
const WebSocket = require("ws");
const WSClient = require("./WSClient");

const AbstractServer = require("./AbstractServer");

/**
 * WebSocket server implementation.
 */
class WSServer extends AbstractServer
{
    constructor(listenOptions)
    {
        super(listenOptions);
        this.wsServer = null;
        this.SocketType = WSClient;
    }

    /**
     * Specifies how the server gets initialized, then creates the server with the specified options.
     */
    _serverCreate()
    {
        const USE_TLS = this.listenOptions.cert != null;

        if(USE_TLS) {
            const tlsOptions = {
                // Server certificate, must be set when using TLS.
                cert: this.listenOptions.cert,
                key: this.listenOptions.key,

                // Enforce client-side certification.
                // If this is set the server will request a client certificate.
                requestCert: this.listenOptions.requestCert,

                // Enforce client connection verification.
                // If this is set to true, then any requested certificate will be validated if requestCert is set.
                rejectUnauthorized: this.listenOptions.rejectUnauthorized,

                // If requesting and validating the client cert (requestCert == true, rejectUnauthorized == true)
                // then we can set the CA cert here which can make a self signed client cert valid in our context.
                ca: this.listenOptions.ca,

                // Milliseconds until we timeout a TLS-handshake.
                // This is important to have when a regular TCP socket is connecting to a TLS socket,
                // and the client is waiting for the server to send first message, then no handshake error will happen because no data is transmitted.
                handshakeTimeout: 30000,
            };
            const https = require("https");
            this.server = https.createServer(tlsOptions);
        }
        else {
            const http = require("http");
            this.server = http.createServer();
        }

        this.server.on("tlsClientError", this._error);
    }

    /**
     * Starts a previously created server listening for connections.
     * Assumes the server is instantiated during object creation.
     */
    _serverListen()
    {
        this.wsServer = new WebSocket.Server({
            path: "/",
            server: this.server,
            noServer: true,
            verifyClient: false,
            clientTracking: true,
            perMessageDeflate: false,
            maxPayload: 100 * 1024 * 1024,
        });

        this.wsServer.on("connection", this._connection);
        this.wsServer.on("error", this._error);
        this.wsServer.on("close", this._close);

        this.server.listen({
            host: this.listenOptions.host,
            port: this.listenOptions.port,
            ipv6Only: this.listenOptions.ipv6Only,
        });
    }

    /**
     * Overrides server close procedure.
     */
    _serverClose()
    {
        if (this.wsServer) {
            this.wsServer.close();
        }
        this.server.close();
    }
}

module.exports = WSServer;
